jax/scipy/linalg.py

```diff
@@ -379,3 +379,196 @@ def _pade13(A):
     V = np.dot(A6, b[12]*A6 + b[10]*A4 + b[8]*A2) + b[6]*A6 + b[4]*A4 + b[2]*A2 + b[0]*ident
     return U,V
+
+@_wraps(scipy.linalg.expm_frechet,lax_description=textwrap.dedent("""\
+    Does not currently support the Scipy argument ``jax.numpy.asarray_chkfinite``, because `jax.numpy.asarray_chkfinite` does not exist at the moment. Does not support the ``method='blockEnlarge'`` argument.
+    """))
+def expm_frechet(A, E, *, method=None, compute_expm=True):
+    return _expm_frechet(A, E, method, compute_expm)
+
+def _expm_frechet(A, E, method=None, compute_expm=True):
+    A = np.asarray(A)
+    E = np.asarray(E)
+    if A.ndim != 2 or A.shape[0] != A.shape[1]:
+        raise ValueError('expected A to be a square matrix')
+    if E.ndim != 2 or E.shape[0] != E.shape[1]:
+        raise ValueError('expected E to be a square matrix')
+    if A.shape != E.shape:
+        raise ValueError('expected A and E to be the same shape')
+    if method is None:
+        method = 'SPS'
+    if method == 'SPS':
+        expm_A, expm_frechet_AE = expm_frechet_algo_64(A, E)
+    else:
+        raise ValueError('only method=\'SPS\' is supported')
+    expm_A, expm_frechet_AE = expm_frechet_algo_64(A, E)
+    if compute_expm:
+        return expm_A, expm_frechet_AE
+    else:
+        return expm_frechet_AE
+
+"""
+Maximal values ell_m of ||2**-s A|| such that the backward error bound
+does not exceed 2**-53.
+"""
+ell_table_61 = (
+        None,
+        # 1
+        2.11e-8,
+        3.56e-4,
+        1.08e-2,
+        6.49e-2,
+        2.00e-1,
+        4.37e-1,
+        7.83e-1,
+        1.23e0,
+        1.78e0,
+        2.42e0,
+        # 11
+        3.13e0,
+        3.90e0,
+        4.74e0,
+        5.63e0,
+        6.56e0,
+        7.52e0,
+        8.53e0,
+        9.56e0,
+        1.06e1,
+        1.17e1,
+        )
+        
+@jit
+def expm_frechet_algo_64(A, E):
+    ident = np.eye(*A.shape, dtype=A.dtype)
+    A_norm_1 = np_linalg.norm(A, 1)
+    """
+    Subset of the Maximal values ell_m of ||2**-s A||
+    such that the backward error bound does not exceed 2**-53.
+    """
+    #ell_table_61 = np.array([1.08e-2,2.00e-1,7.83e-1,1.78e0])
+    ell_table_61_local = np.array([ell_table_61[3],ell_table_61[5],ell_table_61[7],ell_table_61[9]])
+    U3, V3, Lu3, Lv3, s3 = _diff_pade3(A, E, ident)
+    U5, V5, Lu5, Lv5, s5 = _diff_pade5(A, E, ident)
+    U7, V7, Lu7, Lv7, s7 = _diff_pade7(A, E, ident)
+    U9, V9, Lu9, Lv9, s9 = _diff_pade9(A, E, ident)
+    # scaling
+    U13, V13, Lu13, Lv13, s13 = _diff_pade13(A, E, ident, A_norm_1)
+    U = np.select((A_norm_1<=ell_table_61_local),(U3,U5,U7,U9),U13)
+    V = np.select((A_norm_1<=ell_table_61_local),(V3,V5,V7,V9),V13)
+    Lu = np.select((A_norm_1<=ell_table_61_local),(Lu3,Lu5,Lu7,Lu9),Lu13)
+    Lv = np.select((A_norm_1<=ell_table_61_local),(Lv3,Lv5,Lv7,Lv9),Lv13)
+    s = np.select((A_norm_1<=ell_table_61_local),(s3,s5,s7,s9),s13)
```

Comment:
Something to consider here, which is likely a consideration for the `expm` as well, is that this is actually computing all orders of Pade approximations, and then throwing away all but one. (Note that `A_norm_1` is typically a scalar.)

This suggests to me that we should probably be considering one of two approaches here:
1. Use `lax.cond` instead of `np.select`/`np.where`, to avoid computing unneeded terms. This works pretty well on CPUs but unfortunately accelerators are less amenable to control flows.
2. If higher order Pade approximations are strictly better (but more expensive and sometimes unnecessary), then it might be better to _always_ use them, and not bother making the low order approximation.

I don't inherently know which of (1) and (2) is better -- it might depend on the details of the hardware. It might be useful to made some measurements.

Reply:
These are interesting points to consider!

The theory does not explicitly discuss using a higher Pade approximation. So, I cannot say definitively that using a higher Pade approximation is strictly better.

The original implementation by Al-Mohy in [Matlab,](http://eprints.maths.manchester.ac.uk/1442/3/expm_new.zip) `scipy.sparse.linalg.expm`, `scipy.linalg.expm_frechet` all use the appropriate Pade approximation.  `tf.linalg.expm` computes all the approximations, before selecting the appropriate one. 

Using only pade7 for dtype=float32 and dtype=complex64, and only pade13 or dtype=float64 and dtype=complex128 passes _CheckAgainstNumpy. I am not sure if that is necessarily instructive.

Because tf computes all the approximations, I am not sure that there is a better approach available. I can modify the implementation to use lax.cond as you suggested.

Reply:
There are two possible implementations with `lax.cond`.
1. Recursive version that requires `_diff_pade13` to be called each time because of type mismatch in the value of `s` returned by `_diff_pade3`,..`_diff_pade9` and `_diff_pade13`. I have used this approach.
2. Non-recursive version that requires only the appropriate Pade approximation to be executed. This option is, however, not jittable because of the boolean `and` in  `lax.cond`'s `pred` argument.
```
    args = (A, E, ident, A_norm_1)
    U, V, Lu, Lv, s = _diff_pade3(args)
    res = (U, V, Lu, Lv, s)
    
    U, V, Lu, Lv, s = lax.cond(ell_table_61[3]<A_norm_1 and A_norm_1<=ell_table_61[5],
          args, lambda args: _diff_pade5(args),
          res, lambda res: res)
    
    U, V, Lu, Lv, s = lax.cond(ell_table_61[5]<A_norm_1 and A_norm_1<=ell_table_61[7],
          args, lambda args: _diff_pade7(args),
          res, lambda res: res)
        
    U, V, Lu, Lv, s = lax.cond(ell_table_61[7]<A_norm_1 and A_norm_1<=ell_table_61[9],
          args, lambda args: _diff_pade9(args),
          res, lambda res: res)
    
    U, V, Lu, Lv, s = lax.cond(ell_table_61[9]<A_norm_1 ,
          args, lambda args: _diff_pade13(args),
          res, lambda res: res)
```

Reply:
Does it work to use `&` instead of `and`?

Reply:
Ah yes. Replacing `&` by `and` is jittable. Thank you for suggesting it. 

The non-recursive version however also requires a `select` statement for `_diff_pade13(args)` because the following code throws a type mismatch exception for the argument s (the code in the comment above was incorrect):
```
    U, V, Lu, Lv, s = lax.cond(ell_table_61[9]<A_norm_1 ,
          args, lambda args: _diff_pade13(args),
          res, lambda res: res)
```
This means the that the non-recursive version always computes `_diff_pade3` and `_diff_pade13` whereas the recursive version always computes one of `_diff_pade3`, `_diff_padet5`, and_`diff_pade9` as well as `_diff_pade13`. The non-recusive version however performs more conditional expressions. So the recursive version is ultimately preferable.